usvm-jvm/src/main/kotlin/org/usvm/machine/JcDebugProfileObserver.kt

```diff
@@ -0,0 +1,187 @@
+            allNodeStats.values.sum()
+        } else {
+            allNodeStats[slice] ?: 0
+        }
+
+        val forkStats = if (slice == null) {
+            allForkStats.values.sum()
+        } else {
+            allForkStats[slice] ?: 0
+        }
+
+        val selfStat = nodeStats - children.values.sumOf { it.total }
+        val selfForks = forkStats - children.values.sumOf { it.totalForks }
+        return ProfileFrame(inst, nodeStats, selfStat, forkStats, selfForks, children)
+    }
+
+    private class ProfileFrame(
+        val inst: JcInst?,
+        val total: Int,
+        val self: Int,
+        val totalForks: Int,
+        val selfForks: Int,
+        val children: Map<JcInst, ProfileFrame>
+    ) {
+        fun print(str: StringBuilder, indent: String) {
+            val sortedChildren = children.entries
+                .groupBy { it.key.location.method }.entries
+                .sortedByDescending { it.value.sumOf { it.value.total } }
+
+            for ((method, inst) in sortedChildren) {
+                val total = inst.sumOf { it.value.total }
+                val self = inst.sumOf { it.value.self }
+                val totalForks = inst.sumOf { it.value.totalForks }
+                val selfForks = inst.sumOf { it.value.selfForks }
+                str.appendLine("$indent|__ $method | States ${self}/${total} | Forks ${selfForks}/${totalForks}")
+                for ((i, child) in inst.sortedBy { it.value.inst?.location?.index }) {
+                    str.appendLine("$indent$INDENT$i | States ${child.self}/${child.total} | Forks ${child.selfForks}/${child.totalForks}")
+                    child.print(str, "$indent$INDENT$INDENT")
+                }
+            }
+        }
+
+        override fun toString(): String = buildString { print(this, "") }
+
+        companion object {
+            private val INDENT = " ".repeat(4)
+        }
+    }
+
+    override fun onMachineStopped() {
+        logger.debug { report() }
+    }
+
+    private fun report(): String = buildString {
+        reportUnprocessedStates()
+        reportMethodStats()
+        reportProfile()
+    }
+
+    private fun StringBuilder.reportUnprocessedStates() {
+        val unprocessedStates = collectUnprocessedStates()
+        val unprocessedMethods = unprocessedStates.groupBy { it.entrypoint }.mapValues { it.value.size }
+
+        appendLine("Unprocessed states: ${unprocessedStates.size}")
+        unprocessedMethods.entries.sortedByDescending { it.key.enclosingClass.name }.forEach {
+            appendLine("${it.key} | ${it.value}")
+        }
+    }
+
+    private fun StringBuilder.reportMethodStats() {
+        val methodStats = aggregateStats()
+
+        appendLine("Method stats:")
+        for ((cls, methods) in methodStats) {
+            appendLine("-".repeat(20))
+            appendLine("$cls")
+            for ((method, cnt) in methods) {
+                appendLine("$method | $cnt | ${methodCalls[method]?.values?.sum()}")
```

Comment:
Is it possible that we don't have the method in the map? If not, probably it is better to use `getValue`, otherwise, i don't think `null` is a good thing to show here